openvdb_points/unittest/TestIndexIterator.cc

```diff
@@ -0,0 +1,433 @@
+
+        old = ++iter;
+        CPPUNIT_ASSERT_EQUAL(*old, Index32(2));
+        CPPUNIT_ASSERT_EQUAL(*iter, Index32(2));
+
+        CPPUNIT_ASSERT(!iter.next());
+    }
+
+    { // index iterator postfix ++
+        IndexIter iter(0, 3);
+        IndexIter old;
+
+        CPPUNIT_ASSERT(iter);
+        CPPUNIT_ASSERT_EQUAL(*iter, Index32(0));
+
+        old = iter++;
+        CPPUNIT_ASSERT_EQUAL(*old, Index32(0));
+        CPPUNIT_ASSERT_EQUAL(*iter, Index32(1));
+
+        old = iter++;
+        CPPUNIT_ASSERT_EQUAL(*old, Index32(1));
+        CPPUNIT_ASSERT_EQUAL(*iter, Index32(2));
+
+        CPPUNIT_ASSERT(!iter.next());
+    }
+
+    { // overflow iterator ++
+        IndexIter iter;
+
+        iter++;
+        iter++;
+
+        CPPUNIT_ASSERT(!iter);
+    }
+
+    { // index iterator equality
+        IndexIter iter(0, 5);
+
+        CPPUNIT_ASSERT(iter == IndexIter(0, 3));
+        CPPUNIT_ASSERT(iter != IndexIter(1, 3));
+        CPPUNIT_ASSERT(iter == IndexIter(0, 4));
+    }
+}
+
+template <typename IteratorT>
+int count(IteratorT& iter)
+{
+    int total = 0;
+    for (; iter; ++iter) {
+        total++;
+    }
+    return total;
+}
+
+void
+TestIndexIterator::testValueIndexIterator()
+{
+    using namespace openvdb;
+    using namespace openvdb::tree;
+
+    typedef LeafNode<unsigned, 1> LeafNode;
+    typedef LeafNode::ValueOnIter ValueOnIter;
+
+    const int size = LeafNode::SIZE;
+
+    { // one per voxel offset, all active
+        LeafNode leafNode;
+
+        for (int i = 0; i < size; i++) {
+            leafNode.setValueOn(i, i+1);
+        }
+
+        ValueOnIter valueIter = leafNode.beginValueOn();
+
+        IndexValueIter<ValueOnIter> iter(valueIter);
+
+        CPPUNIT_ASSERT(iter);
+
+        CPPUNIT_ASSERT_EQUAL(count(iter), size);
+    }
+
+    { // one per even voxel offsets, only these active
+        LeafNode leafNode;
+
+        int offset = 0;
+
+        for (int i = 0; i < size; i++)
+        {
+            if ((i % 2) == 0) {
+                leafNode.setValueOn(i, ++offset);
+            }
+            else {
+                leafNode.setValueOff(i, offset);
+            }
+        }
+
+        {
+            ValueOnIter valueIter = leafNode.beginValueOn();
+
+            IndexValueIter<ValueOnIter> iter(valueIter);
+
+            CPPUNIT_ASSERT(iter);
+
+            CPPUNIT_ASSERT_EQUAL(count(iter), size/2);
+        }
+    }
+
+    { // one per odd voxel offsets, all active
+        LeafNode leafNode;
+
+        int offset = 0;
+
+        for (int i = 0; i < size; i++)
+        {
+            if ((i % 2) == 1) {
+                leafNode.setValueOn(i, offset++);
+            }
+            else {
+                leafNode.setValueOn(i, offset);
+            }
+        }
+
+        {
+            ValueOnIter valueIter = leafNode.beginValueOn();
+
+            IndexValueIter<ValueOnIter> iter(valueIter);
+
+            CPPUNIT_ASSERT(iter);
+
+            CPPUNIT_ASSERT_EQUAL(count(iter), 3);
```

Comment:
DIM is 1, hence SIZE is 2^1 (8 voxels)

With the above filter, the values become:

0 0 1 1 2 2 3 3

Hence points live in 3rd, 5th, 7th voxel.